pkg/store/bucket.go

```diff
@@ -2303,32 +2351,59 @@ func (r *bucketChunkReader) loadChunks(ctx context.Context, offs []uint32, seq i
 		}
 	}()
 
-	r.chunkBytes = append(r.chunkBytes, b)
 	r.stats.chunksFetchCount++
-	r.stats.chunksFetched += len(offs)
+	r.stats.chunksFetched += len(pIdxs)
 	r.stats.chunksFetchDurationSum += time.Since(fetchBegin)
-	r.stats.chunksFetchedSizeSum += int(end - start)
+	r.stats.chunksFetchedSizeSum += int(part.End - part.Start)
 
-	readOffset := 0
-	for idx, o := range offs {
-		chunkRange := chunkRanges[idx]
+	var (
+		buf        = make([]byte, EstimatedMaxChunkSize)
+		readOffset = int(pIdxs[0].off)
+
+		// Save a few allocations.
+		written  int64
+		diff     uint32
+		chunkLen int
+		n        int
+	)
 
-		// The chunks byte ranges are stored contiguously in the data buffer.
-		cb := (*b)[readOffset : readOffset+chunkRange.length]
-		readOffset += chunkRange.length
+	for i, pIdx := range pIdxs {
+		// Fast forward range reader to the next chunk start in case of sparse (for our purposes) byte range.
+		for readOffset < int(pIdx.off) {
+			written, err = io.CopyN(ioutil.Discard, bufReader, int64(pIdx.off)-int64(readOffset))
+			if err != nil {
+				return errors.Wrap(err, "fast forward range reader")
+			}
+			readOffset += int(written)
+		}
+		// Presume chunk length.
+		chunkLen = EstimatedMaxChunkSize
+		if i+1 < len(pIdxs) {
+			if diff = pIdxs[i+1].off - pIdx.off; int(diff) < chunkLen {
+				chunkLen = int(diff)
+			}
+		}
+		cb := buf[:chunkLen]
+		n, err = io.ReadFull(bufReader, cb)
+		readOffset += n
+		if errors.Is(err, io.ErrUnexpectedEOF) && i == len(pIdxs)-1 {
+			// This may be a valid case.
+		} else if err != nil {
+			return errors.Wrapf(err, "read range for seq %d offset %x", seq, pIdx.off)
+		}
 
 		l, n := binary.Uvarint(cb)
```

Comment:
Let's try to name variables a bit more explicitly. I don't propose java length, but at least `chunkLen`. Since we have it, why not literally doing `chunkLen, n = binary.Uvarint(cb)` and then add necessary missing items? Just want to improve a bit of readability here! WDYT?

Reply:
Nice idea. Changed l to chunkDataLen (there is ChunkLen already, and it's several bytes larger)